nixos/acme: lego run when account is missing

m1cr0man opened the pull request:
###### Motivation for this change

Closes #101445

It is possible that the wrong lego command will be issued if no account credentials exist. This can happen if you follow the new instructions to resolve JWS verification errors.

Annoyingly, I did not include the `email` option in the hash for the accounts directory. If I were to add it now, I would trigger a full certificate renewal + account creation for every certificated generated from this module :wink: Thus I will apply Irish "It'll be grand" logic and not change it now.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

m1cr0man committed:
nixos/acme: lego run whenen account is missing

m1cr0man commented:
Ping ye olde @NixOS/acme

andir commented:
I wonder if we should use stateVersion to decide if we should use the "compatible" mode or the "proper" mode to handle this. IMHO we should carry a proper solution in the source tree.

m1cr0man commented:
That sounds like a good solution. I'll add the conditional logic to cover that today.

m1cr0man committed:
nixos/acme: Add data.email to othersHash in nixos > 20.09

andir commented on nixos/modules/security/acme.nix:
```diff
@@ -104,7 +104,12 @@ let
+    othersHash = mkHash (
+      "${toString acmeServer} ${data.keyType}"
+      + (
+        optionalString (versionOlder "20.09" config.system.stateVersion) data.email
```
This looks good. I just had a thought that might also apply in general in these stateVersion checks.

Should we, at some point, throw a warning that a user should migrate (at their pace within a release) so we can realistically remove backwards compatibility at some point?

flokli replied:
@m1cr0man any opinions on that?

m1cr0man replied:
So in the newer PR, I actually done away with this logic. This was written under the assumption that there was still rate limit issues with multiple certs being generated, but that's not the case now.

m1cr0man replied:
Oh also worth noting I added a note on this matter to the [release notes](https://github.com/NixOS/nixpkgs/pull/106857/files#diff-e5a39d8376b580ef7bb2eb1a17900ed609b638aa453fb04f05569bfff04a0173R263).

flokli replied:
So should this PR be changed before we merge it as requested in https://github.com/NixOS/nixpkgs/pull/101482#issuecomment-745576658? Or do you want to close this one and incorporate it in the other?

m1cr0man replied:
I will make the necessary changes to this PR so it can be merged. I worry it might take too long for the other PR to be merged.

vincentbernat commented:
I have hit the bug fixed by this PR. Any chance it gets merged?

m1cr0man commented:
Doh! I totally forgot this PR existed. I've duplicated some of this work in #106857 . Can we get this merged so I can deal with the merge conflicts?

m1cr0man committed:
nixos/acme: Remove dependency on system version for hash

This means that all systems running from master will trigger
new certificate creation on next rebuild. Race conditions around
multiple account creation are fixed in #106857, not this commit.

symphorien commented:
Do the maintainers think this is fit for a backport to 20.09 ?

m1cr0man commented:
I was going to propose the back port waits until #106857 is ready, as this PR alone does not fix account creation race conditions. However, it could be argued that has always been an issue so there's no harm in back porting it now. I'll let someone else cast the deciding vote.

nh2 commented:
> However, it could be argued that has always been an issue so there's no harm in back porting it now. I'll let someone else cast the deciding vote.

@m1cr0man Please do backport it, I just found this to block me in the middle of a server migration, and I had to apply these patches to fix it.

m1cr0man commented:
Ok, the backport is here: https://github.com/NixOS/nixpkgs/pull/107769/files
